### PR TITLE
ADBDEV-4059-2 Avoid rescan of modifying operations inside the correlated SubPlans.

### PR DESCRIPTION
**Pull request by bimboterminator1**
When a query had a modifying command inside the correlated SubPlan, the
ModifyTable node could be rescanned for each outer tuple. That led to
execution errors (rescan of specific nodes is not supported). This happened
because the ParallelizeCorrelatedSubplanMutator function did not expect the
ModifyTable node inside the correlated SubPlans.

This patch adds the handling of ModifyTable node for correlated SubPlans. The
ParallelizeCorrelatedSubplanMutator function is extended with the ModifyTable
code branch, that simply broadcasts or focuses the result of ModifyTable,
depending on the target flow type. Then the result is materialized in order
to avoid rescan of underlying nodes.

### CODE REVIEW

**Comment by BenderArenadata** [comment]
Allure report https://allure-ee.adsw.io/launch/52693

**Comment by BenderArenadata** [comment]
Allure report

**Review by RekGRpth — DISMISSED** [review]

**Comment by BenderArenadata** [comment]
Allure report https://allure-ee.adsw.io/launch/53169

**Comment by InnerLife0** [comment]
Is there any chance that we can get the same condition as [here](https://github.com/arenadata/gpdb/blob/88f80868d937f42532a6842eb60f71b0035ab67a/src/backend/cdb/cdbllize.c#L578)?

**Comment by bimboterminator1** [comment]
> Is there any chance that we can get the same condition as [here](https://github.com/arenadata/gpdb/blob/88f80868d937f42532a6842eb60f71b0035ab67a/src/backend/cdb/cdbllize.c#L578)?

Let's start from the following query:
```
create table t1 (i int, j int);
create table t2 (i int, j int);
create table t_repl(i int, j int) distributed replicated;
explain (costs off, verbose)
WITH cte AS (INSERT INTO t1 (SELECT i, 10*random() FROM t_repl WHERE t_repl.i > 1 *random()) returning *)
SELECT * FROM t2 WHERE t2.i IN (SELECT cte.i FROM cte WHERE cte.j = t2.i);
explain (costs off)
WITH cte AS (INSERT INTO t1 (SELECT i, 10*random() FROM t_repl WHERE t_repl.i > 1 *random()) returning *)
SELECT * FROM t2 WHERE t2.i IN (SELECT cte.i FROM cte WHERE cte.j = t2.i);
                                                           QUERY PLAN                                                           
--------------------------------------------------------------------------------------------------------------------------------
 Gather Motion 3:1  (slice3; segments: 3)
   ->  Seq Scan on t2
         Filter: (SubPlan 1)
         SubPlan 1  (slice3; segments: 3)
           ->  Subquery Scan on cte
                 Filter: (cte.j = t2.i)
                 ->  Materialize
                       ->  Broadcast Motion 3:3  (slice2; segments: 3)
                             ->  Insert on t1
                                   ->  Redistribute Motion 1:3  (slice1; segments: 1)
                                         Hash Key: "*SELECT*".i
                                         ->  Subquery Scan on "*SELECT*"
                                               ->  Result
                                                     ->  Seq Scan on t_repl
                                                           Filter: ((i)::double precision > ('1'::double precision * random()))
 Optimizer: Postgres query optimizer
```
This query contains volatile quals inside the insert subquery, and replicated distribution is handled as for any common query with volatility. When we are stepping in [here](https://github.com/arenadata/gpdb/blob/fcfac650cc8349ece9662dfeea6a34071f8d0db7/src/backend/optimizer/path/allpaths.c#L601), we get [here](https://github.com/arenadata/gpdb/blob/fcfac650cc8349ece9662dfeea6a34071f8d0db7/src/backend/cdb/cdbpath.c#L2098-L2108), where the proper handling is performed.

Let's consider cases related to this [patch](https://github.com/arenadata/gpdb/commit/cc35273a7a3d2e5c7c69bde151560601aa76a857#diff-4f0575423ee0e1a3434c5af37bdc873165983c0884489e5ae2c71250c66a0ec3).
```
explain (costs off)
WITH cte AS (INSERT INTO t1 (SELECT * from t3 where t3.i IN (select t_repl.i * random() FROM t_repl WHERE t_repl.i > 1 *random() and t_repl.i=t3.j)) returning *)
SELECT * FROM t2 WHERE t2.i IN (SELECT cte.i FROM cte WHERE cte.j = t2.i);
                                                                  QUERY PLAN                                                                  
----------------------------------------------------------------------------------------------------------------------------------------------
 Gather Motion 3:1  (slice3; segments: 3)
   ->  Seq Scan on t2
         Filter: (SubPlan 2)
         SubPlan 2  (slice3; segments: 3)
           ->  Subquery Scan on cte
                 Filter: (cte.j = t2.i)
                 ->  Materialize
                       ->  Broadcast Motion 3:3  (slice2; segments: 3)
                             ->  Insert on t1
                                   ->  Seq Scan on t3
                                         Filter: (SubPlan 1)
                                         SubPlan 1  (slice2; segments: 3)
                                           ->  Result
                                                 ->  Result
                                                       Filter: (t_repl.i = t3.j)
                                                       ->  Materialize
                                                             ->  Broadcast Motion 1:3  (slice1; segments: 1)
                                                                   ->  Seq Scan on t_repl
                                                                         Filter: ((i)::double precision > ('1'::double precision * random()))
```
Here, ParallelizeCorrelatedSubPlan is executed twice: for SubPlan 2, then for SubPlan 1. And case with replicated distribution is handled [correctly](https://github.com/arenadata/gpdb/blob/88f80868d937f42532a6842eb60f71b0035ab67a/src/backend/cdb/cdbllize.c#L578).
Case for uncorrelated SubPlan (the same story).
```
explain (costs off)
WITH cte AS (INSERT INTO t1 (SELECT * from t3 where 0 < ALL(select t_repl.i * random() FROM t_repl WHERE t_repl.i > 1 *random())) returning *)
SELECT * FROM t2 WHERE  0 < ALL (SELECT cte.i FROM cte);
                                                               QUERY PLAN                                                               
----------------------------------------------------------------------------------------------------------------------------------------
 Gather Motion 3:1  (slice3; segments: 3)
   ->  Result
         One-Time Filter: (SubPlan 2)
         ->  Seq Scan on t2
         SubPlan 2  (slice3; segments: 3)
           ->  Materialize
                 ->  Broadcast Motion 3:3  (slice2; segments: 3)
                       ->  Subquery Scan on cte
                             ->  Insert on t1
                                   ->  Result
                                         One-Time Filter: (SubPlan 1)
                                         ->  Seq Scan on t3
                                         SubPlan 1  (slice2; segments: 3)
                                           ->  Materialize
                                                 ->  Broadcast Motion 1:3  (slice1; segments: 1)
                                                       ->  Result
                                                             ->  Seq Scan on t_repl
                                                                   Filter: ((i)::double precision > ('1'::double precision * random()))
 Optimizer: Postgres query optimizer
(19 rows)

```

**Comment by RekGRpth** [comment]
How about subquery insert into replicated table with volatile function?

**Comment by bimboterminator1** [comment]
> How about subquery insert into replicated table with volatile function?

Replicated distribution is gonna be considered in ADBDEV-2160

**Review by RekGRpth — DISMISSED** [review]

File: src/backend/cdb/cdbllize.c [hunk 1]
```diff
@@ -659,6 +659,20 @@ ParallelizeCorrelatedSubPlanMutator(Node *node, ParallelizeCorrelatedPlanWalkerC
...
+	if (IsA(node, ModifyTable))
+	{
+		Plan   *mplan = (Plan *) node;
+		if (ctx->movement == MOVEMENT_BROADCAST)
+			broadcastPlan(mplan, false /* stable */ , false /* rescannable */,
+					ctx->currentPlanFlow->numsegments);
+		else
+			focusPlan(mplan, false /* stable */ , false /* rescannable */);
+
+		Plan   *mat = materialize_subplan((PlannerInfo *) ctx->base.node, mplan);
```
Comment (InnerLife0) [inline src/backend/cdb/cdbllize.c]: Ready to approve, but we should move declaration of this variable upper, according to conventions.
The block above (you inspired by) has logical division to sub-blocks, which is also breaks convention, but anyway.

Comment (RekGRpth) [reply src/backend/cdb/cdbllize.c]:
```c
return (Node *) materialize_subplan((PlannerInfo *) ctx->base.node, mplan);
```

Comment (bimboterminator1) [reply src/backend/cdb/cdbllize.c]: done

File: src/backend/cdb/cdbllize.c [hunk 2]
```diff
@@ -659,6 +659,19 @@ ParallelizeCorrelatedSubPlanMutator(Node *node, ParallelizeCorrelatedPlanWalkerC
 		return ParallelizeCorrelatedSubPlanMutator(node, ctx);
 	}
 
+	if (IsA(node, ModifyTable))
+	{
+		Plan   *mplan = (Plan *) node;
```
Comment (RekGRpth) [inline src/backend/cdb/cdbllize.c]: pgindent sugests more ident, see 655 line

Comment (bimboterminator1) [reply src/backend/cdb/cdbllize.c]: done

Comment (RekGRpth) [reply src/backend/cdb/cdbllize.c]: are you apply pgindent to your changes?

Comment (bimboterminator1) [reply src/backend/cdb/cdbllize.c]: Yep, that's it

**Review by RekGRpth — DISMISSED** [review]

**Review by InnerLife0 — DISMISSED** [review]

File: src/test/regress/sql/subselect_gp.sql [hunk 3]
```diff
@@ -1248,3 +1248,124 @@ select count(*) from t1;
...
+create table t1(i int) distributed by (i);
+create table t2(i int) distributed by (i);
+insert into t2 values (1), (2);
+
+explain (costs off)
+with cte as
+(insert into t1
+ select i from generate_series(1, 5) i
+ returning *)
+ select * from t2
```
Comment (andr-sokolov) [inline src/test/regress/sql/subselect_gp.sql]: Remove please the space before "select" here and below. The select should be on the same level as with. It is not a part of with

Comment (bimboterminator1) [reply src/test/regress/sql/subselect_gp.sql]: done

**Review by RekGRpth — APPROVED** [review]

**Pull request merged**